ddtrace/errortracking/handled_exceptions_after_3_12.py

```diff
@@ -0,0 +1,130 @@
+from functools import lru_cache as cached
+import sys
+from types import CodeType
+from types import ModuleType
+from typing import Callable
+
+from ddtrace.errortracking.handled_exceptions_callbacks import _default_datadog_exc_callback
+from ddtrace.errortracking.handled_exceptions_callbacks import _unhandled_exc_datadog_exc_callback
+from ddtrace.internal.compat import Path
+from ddtrace.internal.module import BaseModuleWatchdog
+from ddtrace.internal.packages import filename_to_package  # noqa: F401
+from ddtrace.internal.packages import is_stdlib  # noqa: F401
+from ddtrace.internal.packages import is_third_party  # noqa: F401
+from ddtrace.internal.packages import is_user_code  # noqa: F401
+from ddtrace.settings.errortracking import config
+
+
+INSTRUMENTED_FILE_PATHS = []
+
+
+def create_should_report_exception_optimized(checks: set[str | None]) -> Callable[[str, Path], bool]:
+    """
+    sys.monitoring reports EVERY handled exceptions, including python internal ones.
+    Therefore we need to filter based on the file_name/file_path. If this check is called
+    many times (it is the case), it becomes costly.
+    This function generates a precompiled version of `should_report_exception` that is as fast
+    as static logic.
+    """
+    logic = ""
+    # It means that all exceptions should be reported. We still want to exclude python internals and ddtrace
+    if len(checks) == 0:
+        logic = "'frozen' not in file_name and is_stdlib(file_path) is False and 'ddtrace' not in file_name"
+    elif "modules" in checks:
+        logic = "file_name in INSTRUMENTED_FILE_PATHS"
+    else:
+        logic = "'frozen' not in file_name"
+        conditions = []
+        if "all_user" in checks:
+            conditions.append("is_user_code(file_path)")
+        if "all_third_party" in checks:
+            conditions.append("(is_third_party(file_path) and filename_to_package(file_path).name != 'ddtrace')")
+        logic += f" and ({' or '.join(conditions)})"
+
+    namespace = {}  # type: ignore
+    exec(
```

Comment:
Can we use assembly instead (e.g. https://github.com/DataDog/dd-trace-py/blob/c6ce9edc7ab7f53e9250a73f1aba8af551e5de81/ddtrace/internal/bytecode_injection/__init__.py#L36-L37)?